Migrate document collection

KludgeKML opened the pull request:
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What

Migrate the document_collection schema rendering from government-frontend to Frontend. 

Note that there's no locale copying in this PR, because document collections only need the names of the various document types, and those were copied into the right place in https://github.com/alphagov/frontend/pull/4848

## Why

https://trello.com/c/gCX6e1Kb/674-move-documentcollection-from-government-frontend-to-frontend, [Jira issue PNP-6446](https://gov-uk.atlassian.net/browse/PNP-6446)

## Screenshots?

### Document collection with single notification button (https://www.gov.uk/government/collections/statutory-guidance-schools)

Before | After
------- | -------
![image](https://github.com/user-attachments/assets/d5544123-f394-4d96-bffe-baeec1c69219) | ![image](https://github.com/user-attachments/assets/fa00e880-2cae-4a83-8d09-bb748680f18a)

### Document collection with topic notification link (https://www.gov.uk/government/collections/self-employment-detailed-information)

Before | After
------ | -------
<img width="977" alt="image" src="https://github.com/user-attachments/assets/4e3589c0-dac0-42b4-a5e1-e8e232312f02" /> |  <img width="972" alt="image" src="https://github.com/user-attachments/assets/cc6d5a0c-46fb-41c3-b95a-cf8cba3e770c" />

KludgeKML committed:
feat: Add placeholder route and controller for document_collection

- document collections are all under /government/collections/, so we
  can use simple routing.

KludgeKML committed:
feat: Copy document_collection view from government-frontend

- copies the view exactly as is from government-frontend. Subsequence commits will update hte view to use values from the ContentItem model.

Audit trail:
https://github.com/alphagov/government-frontend/blob/be848ec7c378a8362f3df387e43f2213882fa32e/app/views/content_items/document_collection.html.erb

KludgeKML committed:
test: Add request tests

KludgeKML committed:
test: Copy integration test for document collection

- Copy the file across as is, changes in following commits

Audit trail:
https://github.com/alphagov/government-frontend/blob/a361915b42cad2326fd6aac2531744429aa1b895/test/integration/document_collection_test.rb

KludgeKML committed:
test: convert to empty spec test

KludgeKML committed:
test: convert title/description test

KludgeKML committed:
test: convert metadata tests

- Also add behaves-like shared example for meta tags

KludgeKML committed:
test: Convert body render test

KludgeKML committed:
test: convert test for contents list

KludgeKML committed:
test: Test that empty collections groups don't appear in the content list

KludgeKML committed:
test: it renders each collection group

KludgeKML committed:
test: it renders all collection documents

KludgeKML committed:
feat: Update README and examples config

KludgeKML committed:
lint: Fix lint errors in document collection show template

- Run `bundle exec erb_lint --lint-all --autocorrect` to fix the errors

KludgeKML committed:
fix: replace @content_item in template with content_item

KludgeKML committed:
fix: Update headers in document collection template

- following pattern from other route migrations

KludgeKML committed:
fix: Add meta description to document collection pages

- machine readable metadata tag only adds og:description, we have
  to add the default one.

KludgeKML committed:
fix: update call to subscribe/unsubscribe flash

- partial introduced in https://github.com/alphagov/frontend/pull/4756
- in government-frontend, the heading_and_context[:text] just passes through to the
  content item's title for the document collection presenter (which inherits this from
  the base content item presenter)

Audit trail:
https://github.com/alphagov/government-frontend/blob/04f02c31e99b20732a5265b0a3744825a7dd9308/app/presenters/content_item_presenter.rb#L27

KludgeKML committed:
fix: update heading values

- using the .name format begun in https://github.com/alphagov/frontend/pull/4848

KludgeKML committed:
fix: simplify variable passed to translation partial

KludgeKML committed:
feat: create document_collection model

KludgeKML committed:
fix: explicit variables in call to notice component

- Following the pattern in other migrations. This uses the methods from the Withdrawable concern
  added to the model in a previous commit.

KludgeKML committed:
feat: make document collection updatable

- view is about to require .updated and .initial_publication_date, provided by this concern.

KludgeKML committed:
feat: add #taxonomy_topic_email_override_base_path method

- this is a special links element found in these document types that allows a topic-based email setup.
- it's an array, but government-frontend only ever uses the first element, if one is present.

Audit trail:
https://github.com/alphagov/government-frontend/blob/04f02c31e99b20732a5265b0a3744825a7dd9308/app/presenters/document_collection/signup_link.rb#L7-L9

KludgeKML committed:
feat: add document_collection_presenter

- We'll need the presenter for the next part of the template.

KludgeKML committed:
feat: add show_emails_signup_link? to presenter

- No previous tests, so added.

Audit trail:
https://github.com/alphagov/government-frontend/blob/1d6be124a3274ae38b97a337ca388b8065c1c9ef/app/presenters/document_collection/signup_link.rb#L3C1-L5C8

KludgeKML committed:
feat: add SinglePageNotificationButton concern to document_collections

- needed for the next part of the template, a single page notification button will be presented if a topic signup link isn't.
- uses concern added here: https://github.com/alphagov/frontend/pull/4769

KludgeKML committed:
fix: remove important_metadata block

- There isn't ever any important metadata for document_collections, so this
  block would never have been called. (document collections presenter in
  government-frontend includes the metadata concern, which always returns
  an empty value for important_metadata unless overridden.

Audit trail:
https://github.com/alphagov/government-frontend/blob/1d6be124a3274ae38b97a337ca388b8065c1c9ef/app/presenters/content_item/metadata.rb#L14-L16

KludgeKML committed:
feat: make document_collection political

- add Political concern and tests, this will be used by the shared/historical
  partial

KludgeKML committed:
feat: add headers to document_collections model

- The contents list in a document_collection is a mix of h2 headings from the body
  (extracted at publish time and supplied as a headings element), and the group
  headings of non-empty collection groups. First we add the headers.

KludgeKML committed:
feat: Add collection_groups to document_collection model

- Add a datatype to represent items from the details/collection_groups element of
  the content item. Each CollectionGroup has a body, id and title, and a list of
  documents, mapped from the documents listed in links. This seems like a natural
  thing to map this way (not complex or shared enough for its own model, but useful
  to be able to manipulate like a model rather than as a big hash).
- NB: Wrapping the details/collection_groups dig with a fallback to an empty array
  is important, because sometimes document_collections are presented as document
  links (for instance, at the time of commit...
  https://www.gov.uk/api/content/government/collections/self-employment-detailed-information
  ...points to another document collection. When document collections appear in
  links, they don't contain their collection_groups, so we can't rely on this
  not returning `nil`.

KludgeKML committed:
fix: use #headers_for_contents_list_component in view

KludgeKML committed:
feat: add topic email signup option

- Using the presenter and model methods added in previous commits.
- Was originally in a shared partial, but it was only used by this view, so inline it.

Audit trail:
https://github.com/alphagov/government-frontend/blob/main/app/views/shared/_document_collections_email_signup.html.erb#L1C21-L1C44

KludgeKML committed:
feat: inline document_collection_body

- this isn't a particularly complex partial, and it isn't used
  anywhere else, so inline it. In this commit we just copy it in
  verbatim.

Audit trail:
https://github.com/alphagov/government-frontend/blob/1d6be124a3274ae38b97a337ca388b8065c1c9ef/app/views/content_items/_document_collection_body.html.erb#L4

KludgeKML committed:
fix: replace @content_item with content_item in inlined partial

KludgeKML committed:
lint: lint inlined partial code

KludgeKML committed:
fix: update calls to govspeak component for body

- Following pattern in other migrations.

KludgeKML committed:
feat: add displayable_collection_groups method

- make this a presenter method because it is making presentational
  decisions (excluding groups from showing if they are empty)
- WARNING note that in the original code, withdrawn documents
  were removed after the empty check happened, which seems
  backwards - could lead to a situation where empty content groups
  were rendered or appeared in the contents list if they entirely
  contained withdrawn documents. I believe the new behaviour
  is correct, but we should pay attention to it.

Audit trail:
https://github.com/alphagov/government-frontend/blob/1d6be124a3274ae38b97a337ca388b8065c1c9ef/app/presenters/document_collection_presenter.rb#L17-L26
https://github.com/alphagov/government-frontend/blob/1d6be124a3274ae38b97a337ca388b8065c1c9ef/app/presenters/document_collection_presenter.rb#L87-L91

KludgeKML committed:
feat: make document collection withdrawable

- we've reached the part of the view that requires .withdrawn?, which is provided by the Withdrawable
  model concern. This is already included in all ContentItems, so just add the test for it.

KludgeKML committed:
fix: use displayable_collection_groups in partial

KludgeKML committed:
fix: replace group heading with simple html

- in practise these headings are always h3, so as part of the heading simplification,
  replace the presentational code with simple inline HTML.

  Audit trail:
  https://github.com/alphagov/government-frontend/blob/b7536e22d5f901052b3e3b8d634ff58ab5a329b0/app/presenters/document_collection_presenter.rb#L47-L56

KludgeKML committed:
fix: use group data objects to simplify view

- follow the pattern for calling govspeak component established in other migrations.

KludgeKML committed:
feat: add method to format group for a document_list component

- Extract the code from displayable_collection_groups that filters out withdrawn documents, and
  use that for selecting maps.

Audit Trail:
https://github.com/alphagov/government-frontend/blob/b7536e22d5f901052b3e3b8d634ff58ab5a329b0/app/presenters/document_collection_presenter.rb#L28-L45
https://github.com/alphagov/government-frontend/blob/b7536e22d5f901052b3e3b8d634ff58ab5a329b0/app/presenters/document_collection_presenter.rb#L60-L73
https://github.com/alphagov/government-frontend/blob/b7536e22d5f901052b3e3b8d634ff58ab5a329b0/test/presenters/document_collection_presenter_test.rb#L91-L129
https://github.com/alphagov/government-frontend/blob/b7536e22d5f901052b3e3b8d634ff58ab5a329b0/test/presenters/document_collection_presenter_test.rb#L207-L222

KludgeKML committed:
fix: use group_as_document_list in view

KludgeKML committed:
fix: wrap published_dates values in formatting methods

- Following convention set in previous migrations.

KludgeKML committed:
fix: Update publisher metadata call

- following style for other migrations, use the publisher_metadata partial and explicitly pass
  values to it.
- uses updatable methods added in a previous commit.
- uses contributor_links call from base content_item presenter

KludgeKML committed:
test: with a withdrawn collection

KludgeKML committed:
feat: add page title method with support for withdrawn

- Similarly, we need to support the withdrawn tag in the page title,
  so update following pattern in other migrations

KludgeKML committed:
test: add test for historical/political banner

KludgeKML committed:
test: test content list rendering with empty document groups

- they should render if the body contains h2s, otherwise not.

KludgeKML committed:
test: Copy integration test for document collection email notifications

- In government-frontend this is in a separate directory, but we haven't gone that way yet with
  frontend, so just copy it into a file in the system directory.
- This is a straight copy, to be converted in subsequent commits.

Audit trail:
https://github.com/alphagov/government-frontend/blob/ab690fe3b8f842c292a57ec8a611e734977d02bc/test/integration/document_collection/email_notifications_test.rb

KludgeKML committed:
test: renders a single page notification button when no override is present

KludgeKML committed:
test: convert to empty spec test

KludgeKML committed:
test: renders a signup link when taxonomy topic email override and locale is english

- schema_type method is never used, so remove

KludgeKML committed:
test: renders the button with the account-only endpoint when user is logged in

KludgeKML committed:
test: check no single-page notification button is rendered if page is not in english

KludgeKML committed:
test: check no signup-link is rendered if page is not in english

deborahchua commented on spec/system/document_collection_spec.rb:
```diff
@@ -67,62 +67,57 @@
+        end
+      end
+
+      context "when h2 headers are supplied" do
```
I think we should also [test that nested headings aren't included in the contents list](https://github.com/alphagov/frontend/blob/main/spec/presenter/corporate_information_page_presenter_spec.rb#L15-L38), as Whitehall will automatically extract all headings from the body and will be shown unless [specifically removed](https://github.com/alphagov/frontend/blob/main/app/presenters/corporate_information_page_presenter.rb#L43-L47).

KludgeKML replied:
It was tested implicitly, in fact, because there are L3 headers in the document_collection_with_body example, but I've made it more explicit https://github.com/alphagov/frontend/pull/4839/files#diff-50919a8b21f824666a72e806f02b0ff691f7050bc5064142d7aa5aba1f01c931R106-R119 (as

deborahchua reviewed (commented):
This is looking great! 👏 I think you're now in the lead for the most number of commits in a consolidation PR 🥇 The atomic commits really helped me review easily, and find my place again when there were changes still being pushed up.

I've left a couple of comments mainly about the contents list headers.

AshGDS reviewed (commented):
Frontend looks OK to me 👍

KludgeKML committed:
feat: add headers_for_contents_list_component to presenter

- Mostly Following the pattern from existing migrations. (there is a future
  task to refactor/DRY up this code, but for the moment this matches - with the exception
  of the method below - what we're doing elsewhere).
- Note that valid_outline_headers is doing the same job as exclude_nested_headings
  in other presenters, but doesn't delete information (changing the model)

Audit trail:
https://github.com/alphagov/frontend/blob/main/app/presenters/corporate_information_page_presenter.rb#L43-L47

KludgeKML committed:
feat: add displayable document groups to the headers returned to the contents list

- up to here we were only displaying h2s from the body (if present), update this
  to include the collection groups headers. We do this by adding them into
  the ContentsOutline and letting the existing presenter do its work.

KludgeKML committed:
feat: add single_page_notification_button to template

- we need to explicitly set skip_account if the user isn't logged in, document collections are the
  one example where we're not supposed to push people to the account only endpoint unless
  they already have an account.

Audit trail:
https://github.com/alphagov/government-frontend/blob/1d6be124a3274ae38b97a337ca388b8065c1c9ef/app/views/shared/_document_collections_email_signup.html.erb#L13C1-L17C9

deborahchua reviewed (approved):
LGTM!
